XRTK-Core/Packages/com.xrtk.core/Providers/Controllers/Hands/BaseHandController.cs

```diff
@@ -0,0 +1,139 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using UnityEngine;
+using XRTK.Definitions.Devices;
+using XRTK.Definitions.Utilities;
+using XRTK.Interfaces.InputSystem;
+using XRTK.Interfaces.Providers.Controllers;
+
+namespace XRTK.Providers.Controllers.Hands
+{
+    /// <summary>
+    /// Base implementation with shared / common elements between all supported hand controllers.
+    /// </summary>
+    public abstract class BaseHandController : BaseController, IMixedRealityHandController
+    {
+        private const float currentVelocityWeight = .8f;
+        private const float newVelocityWeight = .2f;
+
+        // Velocity internal states
+        private float deltaTimeStart;
+        private Vector3 lastPosition;
+        private Vector3 lastPalmNormal;
+        private readonly int velocityUpdateInterval = 9;
+        private int frameOn = 0;
+
+        // Hand ray
+        protected HandRay HandRay { get; } = new HandRay();
```

Comment:
I don't think you should lock HandRay to this type here.

In my implementation, the constructor for BaseHandController takes in a parameterized IHandRay, and the creator of this class can assign the correct implmentation.

This allows the developer to assign custom hand ray implementations, while also allowing slightly different implementations that suit the vendor platform's particular quirks. 

As an example, the HL2 implementation for instance, relies on the palm direction to determine if the ray can be shown - well the Magic leap doesn't provide a palm direction, and it's tricky to make use of it for that.

Reply:
Thanks for the input Eric. I like the idea. I'll revisit this.

Reply:
Agreed, we should use interfaces instead of these fixed constructs where possible to allow for customisation

Reply:
Why do we even need a specialized type of ray?

Reply:
See @provencher's comment on why it is beneficial to be able to change / configure the used hand ray. See also https://github.com/microsoft/MixedRealityToolkit-Unity/pull/6973